plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DynamicFilteringJdbcSplitManager.java

```diff
@@ -0,0 +1,178 @@
+        {
+            this.transaction = requireNonNull(transaction, "transaction is null");
+            this.session = requireNonNull(session, "session is null");
+            this.table = requireNonNull(table, "table is null");
+            this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+            this.constraint = requireNonNull(constraint, "constraint is null");
+            this.dynamicFilteringTimeoutNanos = (long) getDynamicFilteringWaitTimeout(session).getValue(NANOSECONDS);
+            this.startNanos = System.nanoTime();
+        }
+
+        @Override
+        public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+        {
+            long remainingTimeoutNanos = getRemainingTimeoutNanos();
+            if (remainingTimeoutNanos > 0 && dynamicFilter.isAwaitable()) {
+                log.debug(
+                        "Waiting for dynamic filter (query: %s, table: %s, remaining timeout: %s)",
+                        session.getQueryId(),
+                        table,
+                        succinctNanos(remainingTimeoutNanos));
+                // wait for dynamic filter and yield
+                return dynamicFilter.isBlocked()
+                        .thenApply(ignored -> EMPTY_BATCH)
+                        .completeOnTimeout(EMPTY_BATCH, remainingTimeoutNanos, NANOSECONDS);
+            }
+
+            Duration waitingTime = succinctNanos(System.nanoTime() - startNanos);
+            log.debug("Enumerating splits (query %s, table: %s, waiting time: %s, awaitable: %s, completed: %s)",
+                    session.getQueryId(),
+                    table,
+                    waitingTime,
+                    dynamicFilter.isAwaitable(),
+                    dynamicFilter.isComplete());
+            stats.processDynamicFilter(dynamicFilter, waitingTime);
+            return getDelegateSplitSource().getNextBatch(maxSize);
+        }
+
+        @Override
+        public void close()
+        {
+            getOptionalDelegateSplitSource().ifPresent(ConnectorSplitSource::close);
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            if (getRemainingTimeoutNanos() > 0 && dynamicFilter.isAwaitable()) {
+                return false;
+            }
+
+            return getDelegateSplitSource().isFinished();
+        }
+
+        private long getRemainingTimeoutNanos()
+        {
+            return dynamicFilteringTimeoutNanos - (System.nanoTime() - startNanos);
+        }
+
+        private synchronized ConnectorSplitSource getDelegateSplitSource()
+        {
+            if (delegateSplitSource.isPresent()) {
+                return delegateSplitSource.get();
+            }
+
+            delegateSplitSource = Optional.of(delegateSplitManager.getSplits(
+                    transaction,
+                    session,
+                    table.intersectedWithConstraint(dynamicFilter.getCurrentPredicate()),
```

Comment:
It would be great to explain why we merge DF back to `JdbcTableHandle` instead of handling DF directly in `io.trino.plugin.jdbc.BaseJdbcClient#getSplits` (IIRC the reason is to make model extendable)

Reply:
I think that's because we have `delegateSplitManager` here and not `JdbcClient` and that's because some connectors don't use `JdbcSplitManager` and have their own split manager for splits generation. I've added comment about it to class javadoc